File: 4.8-bc/go/stage-3-hexagonal/notifications/infra/http/handler.go
package http

import (
	stdhttp "net/http"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/notifications/app"
	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/notifications/domain"
	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-3-hexagonal/shared/httputil"
)

type Handler struct {
	svc *app.Service
}

func NewHandler(svc *app.Service) *Handler {
	return &Handler{svc: svc}
}

func (h *Handler) Register(r chi.Router) {
	r.Post("/notifications/test", h.sendTest)
}

type sendTestReq struct {
	UserID  string `json:"user_id"`
	Channel string `json:"channel"`
	Payload string `json:"payload"`
}

type sendTestResp struct {
	NotificationID uuid.UUID  `json:"notification_id"`
	Channel        string     `json:"channel"`
	SentAt         *time.Time `json:"sent_at"`
}

func (h *Handler) sendTest(w stdhttp.ResponseWriter, r *stdhttp.Request) {
	var req sendTestReq
	if err := httputil.DecodeBody(r, &req); err != nil {
		httputil.WriteJSON(w, stdhttp.StatusBadRequest, map[string]string{"error": "invalid_request"})
		return
	}
	uid, err := uuid.Parse(req.UserID)
	if err != nil {
		httputil.WriteJSON(w, stdhttp.StatusBadRequest, map[string]string{"error": "invalid_user_id"})
		return
	}
	channel := domain.Channel(req.Channel)
	if channel == "" {
		channel = domain.ChannelEmail
	}
	if req.Payload == "" {
		req.Payload = "test notification"
	}

	n, err := h.svc.Send(r.Context(), uid, channel, req.Payload)
	if err != nil {
		httputil.WriteError(w, err)
		return
	}
	httputil.WriteJSON(w, stdhttp.StatusOK, sendTestResp{
		NotificationID: n.ID,
		Channel:        string(n.Channel),
		SentAt:         n.SentAt,
	})
}

File: 4.8-bc/ts/stage-3-hexagonal/src/notifications/infra/http/handler.ts
// HTTP adapter для Notifications BC.
import type { FastifyInstance } from "fastify";

import { writeError } from "../../../shared/httputil.js";
import type { Service } from "../../app/service.js";
import { CHANNEL_EMAIL, type Channel } from "../../domain/notification.js";

interface SendTestReq {
  user_id?: string;
  channel?: string;
  payload?: string;
}

interface SendTestResp {
  notification_id: string;
  channel: string;
  sent_at: Date | null;
}

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

function asChannel(s: string | undefined): Channel {
  if (s === "email" || s === "push" || s === "sms") return s;
  return CHANNEL_EMAIL;
}

export class Handler {
  constructor(private readonly svc: Service) {}

  register(app: FastifyInstance): void {
    app.post<{ Body: SendTestReq }>(
      "/notifications/test",
      async (req, reply) => {
        const body = req.body ?? {};
        if (!body.user_id || !UUID_RE.test(body.user_id)) {
          return reply.code(400).send({ error: "invalid_user_id" });
        }
        const channel = asChannel(body.channel);
        const payload = body.payload || "test notification";
        try {
          const n = await this.svc.send(body.user_id, channel, payload);
          const resp: SendTestResp = {
            notification_id: n.id,
            channel: n.channel,
            sent_at: n.sentAt,
          };
          return reply.code(200).send(resp);
        } catch (err) {
          req.log.error({ err }, "send notification failed");
          return writeError(reply, err);
        }
      },
    );
  }
}
